core/node/node_storage_init/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+            }
+            (None, None) => {
+                tracing::info!("Node has neither genesis L1 batch, nor snapshot recovery info");
+                if self.strategy.snapshot_recovery.is_some() {
+                    InitDecision::SnapshotRecovery
+                } else {
+                    InitDecision::Genesis
+                }
+            }
+        };
+        Ok(decision)
+    }
+
+    /// Initializes the storage for the node.
+    /// After the initialization, the node can safely start operating.
+    pub async fn run(self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let decision = self.decision().await?;
+
+        // Make sure that we have state to work with.
+        match decision {
+            InitDecision::Genesis => {
+                tracing::info!("Performing genesis initialization");
+                self.strategy
+                    .genesis
+                    .initialize_storage(stop_receiver.clone())
+                    .await?;
+            }
+            InitDecision::SnapshotRecovery => {
+                tracing::info!("Performing snapshot recovery initialization");
+                if let Some(recovery) = &self.strategy.snapshot_recovery {
+                    recovery.initialize_storage(stop_receiver.clone()).await?;
+                } else {
+                    anyhow::bail!(
+                        "Snapshot recovery should be performed, but the strategy is not provided"
+                    );
+                }
+            }
+        }
+
+        // Now we may check whether we're in the invalid state and should perform a rollback.
+        if let Some(reverter) = &self.strategy.block_reverter {
+            if let Some(to_batch) = reverter.is_revert_needed().await? {
+                tracing::info!(l1_batch = %to_batch, "State must be rolled back to L1 batch");
+                tracing::info!("Performing the rollback");
+                reverter.revert_storage(to_batch, stop_receiver).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Checks if the node can safely start operating.
+    pub async fn wait_for_initialized_storage(
+        &self,
+        stop_receiver: watch::Receiver<bool>,
+    ) -> anyhow::Result<()> {
+        const POLLING_INTERVAL: Duration = Duration::from_secs(1);
+
+        let decision = self.decision().await?;
+
+        // Wait until data is added to the database.
+        poll(stop_receiver.clone(), POLLING_INTERVAL, || {
+            self.is_database_initialized(decision)
+        })
+        .await?;
+        if *stop_receiver.borrow() {
+            return Ok(());
+        }
+
+        // Wait until the rollback is no longer needed.
+        poll(stop_receiver.clone(), POLLING_INTERVAL, || {
+            self.is_rollback_not_needed(stop_receiver.clone())
+        })
+        .await?;
+
+        Ok(())
+    }
+
+    async fn is_database_initialized(&self, decision: InitDecision) -> anyhow::Result<bool> {
```

Comment:
Not entirely sure, but logically, we should be OK as long as the storage is initialized in any acceptable way. So, roughly speaking, check that either `genesis.is_initialized()` or `strategy.snapshot_recovery.is_initialized()` are true if both are provided. Mostly raising this because `bail!` below looks weird; we don't *perform* snapshot recovery in this method.